File: intime/app/src/main/java/com/vpe_soft/intime/intime/recyclerview/RecyclerViewCursorAdapter.java
package com.vpe_soft.intime.intime.recyclerview;

import android.content.Context;
import android.database.Cursor;
import androidx.cursoradapter.widget.CursorAdapter;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.Locale;

public abstract class RecyclerViewCursorAdapter<T extends RecyclerViewCursorViewHolder> extends RecyclerView.Adapter<T> {

    protected final Context context;
    protected final Locale locale;
    protected CursorAdapter cursorAdapter;
    private T viewHolder;
    protected RecyclerViewCursorAdapter(Context context, Locale locale) {
        this.context = context;
        this.locale = locale;
    }

    protected void setupCursorAdapter(Cursor cursor, int flags, final int resource, final boolean attachToRoot) {
        this.cursorAdapter = new CursorAdapter(context, cursor, flags) {
            @Override
            public View newView(Context context, Cursor cursor, ViewGroup parent) {
                return LayoutInflater.from(context).inflate(resource, parent, attachToRoot);
            }

            @Override
            public void bindView(View view, Context context, Cursor cursor) {
                viewHolder.bindCursor(cursor);
            }
        };
    }

    @Override
    public int getItemCount() {
        return cursorAdapter.getCount();
    }

    public void swapCursor(Cursor cursor) {
        this.cursorAdapter.swapCursor(cursor);
        notifyDataSetChanged();
    }

    protected void setViewHolder(T viewHolder) {
        this.viewHolder = viewHolder;
    }
}

File: intime/app/src/main/java/com/vpe_soft/intime/intime/recyclerview/RecyclerViewCursorAdapter.kt
package com.vpe_soft.intime.intime.recyclerview

import android.content.Context
import android.database.Cursor
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.cursoradapter.widget.CursorAdapter
import androidx.recyclerview.widget.RecyclerView
import com.vpe_soft.intime.intime.activity.MainActivity
import java.util.*

abstract class RecyclerViewCursorAdapter<T : RecyclerViewCursorViewHolder> protected constructor(
    protected val context: Context
) :
    RecyclerView.Adapter<T>() {
    protected var mainActivity: MainActivity = context as MainActivity
    protected lateinit var cursorAdapter: CursorAdapter
    private lateinit var viewHolder: T
    protected fun setupCursorAdapter(
        cursor: Cursor,
        @Suppress("SameParameterValue") resource: Int
    ) {
        cursorAdapter = object : CursorAdapter(context, cursor, 0) {
            override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View =
                LayoutInflater.from(context).inflate(resource, parent, false)

            override fun bindView(view: View?, context: Context, cursor: Cursor) =
                viewHolder.bindCursor(cursor)
        }
    }

    override fun getItemCount(): Int = cursorAdapter.count

    fun swapCursor(cursor: Cursor) {
        cursorAdapter.swapCursor(cursor)
        notifyDataSetChanged()
    }

    protected fun setViewHolder(newViewHolder: T) {
        viewHolder = newViewHolder
    }
}